### PR TITLE
Docker cluster runner

### PR DESCRIPTION
**Pull request by idruzhitskiy**

### CODE REVIEW

**Comment by agend** [comment]
@idruzhitskiy please fix conflicts

File: src/bin/dcr.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        config_dir.to_string(),
+        "/tmp".to_string(),
+    );
+    let configuration = get_configuration()?;
+    configuration.save_cluster_configuration(config_dir)?;
+    let compose = configuration.create_docker_compose(fs_configuration, "bobnet".to_string())?;
+    let arc = Arc::new(Mutex::new(Box::new(compose)));
+    let ctrlc_arc = arc.clone();
+    ctrlc::set_handler(move || {
+        let compose = ctrlc_arc.lock().unwrap();
```
Comment (pyakushin) [inline src/bin/dcr.rs:30]: Is it possible thread lock here?

File: src/bin/dcr.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        compose
+            .down(config_dir)
+            .expect("failed to run \"down\"")
+            .wait_with_output()
+            .expect("failed to wait for compose down otuput");
+    })?;
+    arc.lock().unwrap().up(config_dir)?.wait_with_output()?;
```
Comment (pyakushin) [inline src/bin/dcr.rs:37]: Because we won't release mutex until the end of main() in the same thread (presumably)

File: src/bin/dcr.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    Ok(())
+}
+
+fn get_configuration() -> Result<TestClusterConfiguration, Box<dyn Error>> {
+    let matches = App::new(env!("CARGO_PKG_NAME"))
+        .arg(
+            Arg::with_name("config")
```
Comment (pyakushin) [inline src/bin/dcr.rs:44]: If we have default config in git repo, it makes sense to add it as default arg value here

File: src/bin/dcr.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                .help("config for docker cluster")
+                .takes_value(true)
+                .short("c")
+                .required(true)
+                .long("config"),
+        )
+        .get_matches();
+    let config_filename = matches.value_of("config").expect("required");
+    let file_content = fs::read_to_string(config_filename)?;
+    let configuration: TestClusterConfiguration = serde_yaml::from_str(&file_content)?;
```
Comment (pyakushin) [inline src/bin/dcr.rs:54]: You can deserialize directly from file, with `serde_yaml::from_reader`

File: build.rs [hunk 5]
```diff
@@ -4,6 +4,7 @@ fn main() {
     tonic_build::configure()
         .build_server(true)
         .build_client(true)
+        .format(false)
```
[the remark this answers is not in the document]
Comment (idruzhitskiy) [reply build.rs:7]: We have external format in build scripts, and sometimes rustfmt isn't available for latest nightly, so dockers won't build

File: cluster_test/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# build image
+FROM rustlang/rust:nightly as cargo-build
+
+RUN apt-get update
+
+RUN apt-get install musl-tools -y
+
+RUN rustup target add x86_64-unknown-linux-musl
```
[the remark this answers is not in the document]
Comment (idruzhitskiy) [reply cluster_test/Dockerfile:8]: Yep, for usage in alpine image

File: src/bin/dcr.rs (hunk 1 above)
Comment (idruzhitskiy) [reply src/bin/dcr.rs:30]: Actually no, I'll introduce new var so it will be clear

File: src/bin/dcr.rs (hunk 4 above)
Comment (idruzhitskiy) [reply src/bin/dcr.rs:54]: `    let file = File::open(path)?;
    let reader = BufReader::new(file);
    let obj = serde_yaml::from_reader(reader)?;`

Something like this? Reading as string looks more concise...

File: src/bin/dcr.rs (hunk 3 above)
Comment (idruzhitskiy) [reply src/bin/dcr.rs:45]: Added default value

File: src/bin/dcr.rs (hunk 2 above)
Comment (idruzhitskiy) [reply src/bin/dcr.rs]: Added new variable

File: Cargo.toml [hunk 7]
```diff
@@ -36,6 +36,8 @@ rocket = "0.4.4"
 rocket_contrib = "0.4.4"
 async-trait = "0.1.31"
 termion = "1.5"
+derive-new = "0.5.8"
+ctrlc = "*"
```
[the remark this answers is not in the document]
Comment (idruzhitskiy) [reply Cargo.toml]: Fixed 3.1.4 version

**Comment by agend** [comment]
@idruzhitskiy could you add some documentation on how to use it? that would be useful

**Pull request merged**